### PR TITLE
add ByRow

### PR DESCRIPTION
**Pull request by bkamins**
Following the discussion in https://github.com/JuliaData/Tables.jl/issues/287.

CC @quinnj @nalimilan @pdeffebach

### CODE REVIEW

**Comment by bkamins** [comment]
@quinnj I have also made some small clean-up proposals to src/namedtuples.jl - I hope they are correct 😄.

File: src/utils.jl [hunk 1]
```diff
@@ -160,3 +160,81 @@ rowmerge(row; fields_to_merge...) = rowmerge(row, values(fields_to_merge))
 
 _row_to_named_tuple(row::NamedTuple) = row
 _row_to_named_tuple(row) = NamedTuple(Row(row))
+
+"""
+    ByRow
+
+`ByRow(f)` is a function that is wrapper around function `f` that vectorizes it.
```
Comment (nalimilan) [inline src/utils.jl]: More explicit than "vectorize"?
```suggestion
`ByRow(f)` returns a function which applies function `f` to each element in a vector.
```

File: src/utils.jl [hunk 2]
```diff
@@ -160,3 +160,81 @@ rowmerge(row; fields_to_merge...) = rowmerge(row, values(fields_to_merge))
 
 _row_to_named_tuple(row::NamedTuple) = row
 _row_to_named_tuple(row) = NamedTuple(Row(row))
+
+"""
+    ByRow
```
Comment (nalimilan) [inline src/utils.jl]:
```suggestion
    ByRow <: Function
```

File: src/utils.jl [hunk 3]
```diff
@@ -160,3 +160,81 @@ rowmerge(row; fields_to_merge...) = rowmerge(row, values(fields_to_merge))
...
+
+As a special case `ByRow(f)` also accepts passing a `Tables.ColumnTable` as its
+argument. In this case the function `f` is passed a `NamedTuple` that is created
+from rows of passed table.
+
+The return value of `ByRow` is always a vector.
+
+Vectors passed to `ByRow(f)` must have equal length and use 1-based indexing.
+Also vectors stored in the `Tables.ColumnTable` if it is passed to `ByRow(f)`
+must meet this condition.
```
Comment (nalimilan) [inline src/utils.jl]:
```suggestion
`ByRow(f)` can be passed two types of arguments:
- One or more 1-based `AbstractVector`s of equal length: In this case the returned value
is a vector resulting from applying `f` to elements of passed vectors element-wise. 
Function `f` is called exactly once for each element of passed vectors (as opposed to `map`
which assumes for some types of source vectors (e.g. `SparseVector`) that the
wrapped function is pure, and may call the function `f` only once for multiple
equal values.
- A `Tables.ColumnTable` holding 1-based columns of equal length: In this case the function
`f` is passed a `NamedTuple` created for each row of passed table.

The return value of `ByRow(f)` is always a vector.
```

File: test/runtests.jl [hunk 4]
```diff
@@ -846,3 +846,36 @@ Tables.columnnames(::WideTable2) = [Symbol("x", i) for i = 1:1000]
...
+    @test Tables.ByRow(x -> x^2)(1:3) == [1, 4, 9]
+    @test Tables.ByRow((x, y) -> x*y)(1:3, 2:4) == [2, 6, 12]
+    @test Tables.ByRow((x, y, z) -> x*y*z)(1:3, 2:4, 3:5) == [6, 24, 60]
+    let i = 0
+        @test Tables.ByRow(x -> (i += 1; return i))(spzeros(5)) == 1:5
+    end
+    # check that map is not what we want
+    let i = 0
+        @test map(x -> (i += 1; return i), spzeros(5)) == [1, 1, 1, 1, 1]
+    end
```
Comment (nalimilan) [inline test/runtests.jl:860]: Do we really need to test this? We don't really care how `map` is implemented in SparseArrays.

**Review by nalimilan — APPROVED** [review]

File: test/runtests.jl (hunk 4 above)
Comment (bkamins) [reply test/runtests.jl:860]: We do not need it, but I think it is not a problem and shows the reader of tests the difference (which is not obvious I would say).

File: src/utils.jl [hunk 5]
```diff
@@ -160,3 +160,75 @@ rowmerge(row; fields_to_merge...) = rowmerge(row, values(fields_to_merge))
 
 _row_to_named_tuple(row::NamedTuple) = row
 _row_to_named_tuple(row) = NamedTuple(Row(row))
+
+"""
+    ByRow <: Function
+
+`ByRow(f)` returns a function which applies function `f` to each element in a vector.
+
+`ByRow(f)` can be passed two types of arguments:
```
Comment (bkamins) [inline src/utils.jl:169]: @nalimilan also being 1-based is a requirement.

**Comment by bkamins** [comment]
@quinnj and @pdeffebach - can you please check and approve as this is a surprisingly complex PR (fortunately not big)

File: src/utils.jl [hunk 6]
```diff
@@ -160,3 +160,75 @@ rowmerge(row; fields_to_merge...) = rowmerge(row, values(fields_to_merge))
...
+julia> ByRow((x, y) -> x*y)(1:3, 2:4)
+3-element Vector{Int64}:
+  2
+  6
+ 12
+
+julia> ByRow(x -> x.a)((a=1:2, b=3:4))
+2-element Vector{Int64}:
+ 1
+ 2
```
Comment (quinnj) [inline src/utils.jl:207]:
```suggestion
 2
 
 julia> ByRow(x -> (a=x.a*2, b=sin(x.b), c=x.c))((a=[1,2,3], b=[1.2, 3.4, 5.6], c=["a","b","c"]))
3-element Vector{NamedTuple{(:a, :b, :c), Tuple{Int64, Float64, String}}}:
 (a = 2, b = 0.9320390859672263, c = "a")
 (a = 4, b = -0.2555411020268312, c = "b")
 (a = 6, b = -0.6312666378723216, c = "c")

```

Comment (quinnj) [reply src/utils.jl:207]: I think it'd be nice to have a namedtuple -> namedtuple example here. I tend to prefer in documentation to include package prefix when the item isn't exported; would you mind changing all the usages here to `Tables.ByRow`, so it's extra clear?

**Review by quinnj — APPROVED** [review]
Sorry for the slow review, but this LGTM. I left one comment on an example doc + doc package prefix.

**Comment by bkamins** [comment]
Fixed. The `Tables.` prefix for sure should be there. I had DataFrames.jl loaded, which exports it so I forgotten to add it in the docstring.

**Pull request merged**